// ra-data-opencrud/lib/buildGqlQuery.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildApolloArgs = exports.buildArgs = exports.getArgType = exports.buildFields = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _graphql = require('graphql');

var _raDataGraphql = require('ra-data-graphql');

var _reactAdmin = require('react-admin');

var _gqlTypes = require('./utils/gqlTypes');

var gqlTypes = _interopRequireWildcard(_gqlTypes);

var _getFinalType = require('./utils/getFinalType');

var _getFinalType2 = _interopRequireDefault(_getFinalType);

var _isList = require('./utils/isList');

var _isList2 = _interopRequireDefault(_isList);

var _isRequired = require('./utils/isRequired');

var _isRequired2 = _interopRequireDefault(_isRequired);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var buildFields = exports.buildFields = function buildFields(introspectionResults) {
  return function (fields) {
    return fields.reduce(function (acc, field) {
      var type = (0, _getFinalType2.default)(field.type);

      if (type.name.startsWith('_')) {
        return acc;
      }

      if (type.kind !== _graphql.TypeKind.OBJECT) {
        return [].concat(_toConsumableArray(acc), [gqlTypes.field(gqlTypes.name(field.name))]);
      }

      var linkedResource = introspectionResults.resources.find(function (r) {
        return r.type.name === type.name;
      });

      if (linkedResource) {
        return [].concat(_toConsumableArray(acc), [gqlTypes.field(gqlTypes.name(field.name), {
          selectionSet: gqlTypes.selectionSet([gqlTypes.field(gqlTypes.name('id'))])
        })]);
      }

      var linkedType = introspectionResults.types.find(function (t) {
        return t.name === type.name;
      });

      if (linkedType) {
        return [].concat(_toConsumableArray(acc), [gqlTypes.field(gqlTypes.name(field.name), {
          selectionSet: gqlTypes.selectionSet(buildFields(introspectionResults)(linkedType.fields))
        })]);
      }

      // NOTE: We might have to handle linked types which are not resources but will have to be careful about
      // ending with endless circular dependencies
      return acc;
    }, []);
  };
};

var getArgType = exports.getArgType = function getArgType(arg) {
  var type = (0, _getFinalType2.default)(arg.type);
  var required = (0, _isRequired2.default)(arg.type);
  var list = (0, _isList2.default)(arg.type);

  if (list) {
    if (required) {
      return gqlTypes.listType(gqlTypes.nonNullType(gqlTypes.namedType(gqlTypes.name(type.name))));
    }
    return gqlTypes.listType(gqlTypes.namedType(gqlTypes.name(type.name)));
  }

  if (required) {
    return gqlTypes.nonNullType(gqlTypes.namedType(gqlTypes.name(type.name)));
  }

  return gqlTypes.namedType(gqlTypes.name(type.name));
};

var buildArgs = exports.buildArgs = function buildArgs(query, variables) {
  if (query.args.length === 0) {
    return [];
  }

  var validVariables = Object.keys(variables).filter(function (k) {
    return typeof variables[k] !== 'undefined';
  });
  return query.args.filter(function (arg) {
    return validVariables.includes(arg.name);
  }).reduce(function (acc, arg) {
    return [].concat(_toConsumableArray(acc), [gqlTypes.argument(gqlTypes.name(arg.name), gqlTypes.variable(gqlTypes.name(arg.name)))]);
  }, []);
};

var buildApolloArgs = exports.buildApolloArgs = function buildApolloArgs(query, variables) {
  if (query.args.length === 0) {
    return [];
  }

  var validVariables = Object.keys(variables).filter(function (k) {
    return typeof variables[k] !== 'undefined';
  });

  return query.args.filter(function (arg) {
    return validVariables.includes(arg.name);
  }).reduce(function (acc, arg) {
    return [].concat(_toConsumableArray(acc), [gqlTypes.variableDefinition(gqlTypes.variable(gqlTypes.name(arg.name)), getArgType(arg))]);
  }, []);
};

//TODO: validate fragment against the schema
var buildFieldsFromFragment = function buildFieldsFromFragment(fragment, resourceName, fetchType) {
  var parsedFragment = {};

  if ((typeof fragment === 'undefined' ? 'undefined' : _typeof(fragment)) === 'object' && fragment.kind && fragment.kind === 'Document') {
    parsedFragment = fragment;
  }

  if (typeof fragment === 'string') {
    if (!fragment.startsWith('fragment')) {
      fragment = 'fragment tmp on ' + resourceName + ' ' + fragment;
    }

    try {
      parsedFragment = (0, _graphql.parse)(fragment);
    } catch (e) {
      throw new Error('Invalid fragment given for resource \'' + resourceName + '\' and fetchType \'' + fetchType + '\' (' + e.message + ').');
    }
  }

  return parsedFragment.definitions[0].selectionSet.selections;
};

exports.default = function (introspectionResults) {
  return function (resource, aorFetchType, queryType, variables, fragment) {
    var sortField = variables.sortField,
        sortOrder = variables.sortOrder,
        countVariables = _objectWithoutProperties(variables, ['sortField', 'sortOrder']);

    var apolloArgs = buildApolloArgs(queryType, variables);
    var args = buildArgs(queryType, variables);
    var countArgs = buildArgs(queryType, countVariables);
    var fields = !!fragment ? buildFieldsFromFragment(fragment, resource.type.name, aorFetchType) : buildFields(introspectionResults)(resource.type.fields);

    if (aorFetchType === _reactAdmin.GET_LIST || aorFetchType === _reactAdmin.GET_MANY || aorFetchType === _reactAdmin.GET_MANY_REFERENCE) {
      return gqlTypes.document([gqlTypes.operationDefinition('query', gqlTypes.selectionSet([gqlTypes.field(gqlTypes.name(queryType.name), {
        alias: gqlTypes.name('items'),
        arguments: args,
        selectionSet: gqlTypes.selectionSet(fields)
      }), gqlTypes.field(gqlTypes.name(queryType.name + 'Connection'), {
        alias: gqlTypes.name('total'),
        arguments: countArgs,
        selectionSet: gqlTypes.selectionSet([gqlTypes.field(gqlTypes.name('aggregate'), {
          selectionSet: gqlTypes.selectionSet([gqlTypes.field(gqlTypes.name('count'))])
        })])
      })]), gqlTypes.name(queryType.name), apolloArgs)]);
    }

    if (aorFetchType === _reactAdmin.DELETE) {
      return gqlTypes.document([gqlTypes.operationDefinition('mutation', gqlTypes.selectionSet([gqlTypes.field(gqlTypes.name(queryType.name), {
        alias: gqlTypes.name('data'),
        arguments: args,
        selectionSet: gqlTypes.selectionSet([gqlTypes.field(gqlTypes.name('id'))])
      })]), gqlTypes.name(queryType.name), apolloArgs)]);
    }

    return gqlTypes.document([gqlTypes.operationDefinition(_raDataGraphql.QUERY_TYPES.includes(aorFetchType) ? 'query' : 'mutation', gqlTypes.selectionSet([gqlTypes.field(gqlTypes.name(queryType.name), {
      alias: gqlTypes.name('data'),
      arguments: args,
      selectionSet: gqlTypes.selectionSet(fields)
    })]), gqlTypes.name(queryType.name), apolloArgs)]);
  };
};
