// ra-core/esm/index.js
export * from './dataFetchActions';
export * from './actions';
export * from './auth';
export * from './i18n';
export * from './util';
export * from './controller';
export * from './form';
import _createAppReducer from './reducer';
export { _createAppReducer as createAppReducer };
export { getResources, getReferenceResource, getLocale, getNotification, getPossibleReferences, getPossibleReferenceValues } from './reducer';
import _adminReducer from './reducer/admin';
export { _adminReducer as adminReducer };
import _i18nReducer from './reducer/i18n';
export { _i18nReducer as i18nReducer };
import _queryReducer from './reducer/admin/resource/list/queryReducer';
export { _queryReducer as queryReducer };


export { getIds, getReferences, getReferencesByIds, nameRelatedTo } from './reducer/admin/references/oneToMany';

export * from './sideEffect';
import _CoreAdmin from './CoreAdmin';
export { _CoreAdmin as CoreAdmin };
import _CoreAdminRouter from './CoreAdminRouter';
export { _CoreAdminRouter as CoreAdminRouter };
import _createAdminStore from './createAdminStore';
export { _createAdminStore as createAdminStore };
import _RoutesWithLayout from './RoutesWithLayout';
export { _RoutesWithLayout as RoutesWithLayout };
import _Resource from './Resource';
export { _Resource as Resource };
